HW2/P5/driver.py

```diff
@@ -80,3 +97,23 @@ def randcolor():
             # SUBPROBLEM 3: sort objects by location.  Be sure to update the
             # grid if objects' indices change!  Also be sure to sort the
             # velocities with their object positions!
+            
+            # numbering positions : (1, [x1, y1]), (2, [x2, y2]), ...
+            index_and_position = zip(range(len(positions)), (positions / grid_spacing).astype(int))
+            # sort using Z order curve
+            index_and_position.sort(cmp_zorder)
+            # get new order
+            new_order = map(lambda (i, position): i, index_and_position)
+            # reset grid
+            grid[:, :] = -1
+            
+            # reorder the existing position, velocity 
+            positions = np.array([positions[i] for i in new_order])
+            velocities = np.array([velocities[i] for i in new_order])
+
+
+            # update the grid indices
+            for i in range(num_balls):
+                grid[(positions[i, 0] / grid_spacing).astype(int),
+                     (positions[i, 1] / grid_spacing).astype(int)] = i 
+
```

Comment:
You can use the same code as above to initialize the grid for more Pythonic succinctness.